File: src/ipc/messages/types/structs/Navigation.ts
import * as ldm from '../../leapdatamodel';
import * as valuetype from '../../valuetype';
import type * as types from '..';
import { StructBase as _StructBase } from '../../StructBase';
import { LeapError as _LeapError } from '../../LeapError';
import { ValueBase as _ValueBase } from '../../ValueBase';

export class Navigation extends _StructBase {
    static _type: string = 'navigation';
    static _leapStruct: valuetype.LeapStruct = new valuetype.LeapStruct("navigation", [], [new valuetype.Property("global-id", new valuetype.ValueType("int", [])), new valuetype.Property("path", new valuetype.ValueType("str", []))]);
    globalId: number;
    path: string;

    constructor(globalId: number, path: string, ) {
        super();
        this.globalId = globalId;
        this.path = path;
    }

    _verifyAsTypeOrRaise(path: string[], valueType: valuetype.ValueType): void {
        if (valueType.name == Navigation._type) {
            const structType = Navigation._leapStruct.applyArgs(valueType.args);
            _ValueBase._verifyValueOrRaise(path.concat(['globalId']), this.globalId, structType.props[0].propType);
            _ValueBase._verifyValueOrRaise(path.concat(['path']), this.path, structType.props[1].propType);
        } else {
            throw _ValueBase._pathError(path, `expecting ${Navigation._type}`);
        }
    }

    _toLdm(valueType: valuetype.ValueType): ldm.Value {
        let value = new Map<string, ldm.Value>();
        const structType = Navigation._leapStruct.applyArgs(valueType.args);
        value.set("global-id", _ValueBase._valueToLdm(this.globalId, structType.props[0].propType));
        value.set("path", _ValueBase._valueToLdm(this.path, structType.props[1].propType));
        return new ldm.StructValue(value, valueType);
    }
}

File: src-tauri/src/ipc/messages/types/structs/navigation.rs
use std::collections::HashMap as _HashMap;
#[allow(unused_imports)]
use super::super::super::types;
use super::super::super::value_base::ValueBase as _ValueBase;
use super::super::super::struct_base::StructBase as _StructBase;
use super::super::super::leap_data_model as ldm;
use super::super::super::value_type::ValueType as _ValueType;

#[derive(Clone, Debug)]
pub struct Navigation {
    pub global_id: i64,
    pub path: String,
}

impl Navigation {
    pub fn new(global_id: i64, path: String, ) -> Self {
        Self {
            global_id,
            path,
        }
    }
}

impl _ValueBase for Navigation {
    const _TYPE: &'static str = "navigation";

    fn _applied_args() -> Vec<_ValueType> {
        vec![
        ]
    }

    fn _to_ldm(&self) -> ldm::Value {
        ldm::Value::Struct(self._to_ldm_struct())
    }
}

impl _StructBase for Navigation {
    fn _to_ldm_struct(&self) -> ldm::StructValue {
        let mut value = _HashMap::new();
        value.insert("global-id".to_owned(), self.global_id._to_ldm());
        value.insert("path".to_owned(), self.path._to_ldm());
        ldm::StructValue{
            value,
            value_type: _ValueType::new(Self::_TYPE.to_owned(), Self::_applied_args()),
        }
    }
}